### PR TITLE
ISPN-10854 REST Task API

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-10854

### CODE REVIEW

**Comment by gustavonalle** [comment]
Looks like there are some related failures

File: cli/cli-client/src/main/java/org/infinispan/cli/resources/TasksResource.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+      } else if (getConnection().getAvailableTasks(getParent().getName()).contains(name)) {
+         return new TaskResource(this, name);
+      } else {
+         throw Messages.MSG.noSuchResource(name);
+      }
+   }
+
+   @Override
+   public String describe() throws IOException {
+      return NAME;//TODO
```
Comment (karesti) [inline cli/cli-client/src/main/java/org/infinispan/cli/resources/TasksResource.java]: //TODO ?

File: tasks/manager/src/test/java/org/infinispan/tasks/TaskManagerTest.java [hunk 2]
```diff
@@ -66,9 +66,18 @@ public void testRunTask() throws InterruptedException, ExecutionException {
       assertEquals(EventLogCategory.TASKS, memoryLogger.getCategory());
       assertEquals(EventLogLevel.INFO, memoryLogger.getLevel());
 
+      memoryLogger.reset();
```
Comment (karesti) [inline tasks/manager/src/test/java/org/infinispan/tasks/TaskManagerTest.java]: could not this be in another method ?

File: tasks/manager/src/test/java/org/infinispan/tasks/DummyTaskEngine.java [hunk 3]
```diff
@@ -45,6 +46,9 @@ public String getName() {
       switch (DummyTaskTypes.valueOf(taskName)) {
       case SUCCESSFUL_TASK:
          return (CompletableFuture<T>) CompletableFuture.completedFuture("result");
+      case PARAMETERIZED_TASK:
+         Map<String, ?> params = context.getParameters().get();
```
Comment (karesti) [inline tasks/manager/src/test/java/org/infinispan/tasks/DummyTaskEngine.java]: what if is Optional.empty ? In any case, for me this is an anti-pattern.
We should not be using an Optional<Map> this way but an empty map IMO

File: server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+      return new Invocations.Builder()
+            .invocation().methods(GET).path("/v2/tasks/").handleWith(this::listTasks)
+            .invocation().methods(PUT, POST).path("/v2/tasks/{taskName}").handleWith(this::createScriptTask)
+            .invocation().methods(GET).path("/v2/tasks/{taskName}").withAction("exec").handleWith(this::runTask)
+            .create();
+   }
+
+   private CompletionStage<RestResponse> listTasks(RestRequest request) {
+      EmbeddedCacheManager cacheManager = invocationHelper.getRestCacheManager().getInstance();
+      TaskManager taskManager = SecurityActions.getGlobalComponentRegistry(cacheManager).getComponent(TaskManager.class);
```
Comment (karesti) [inline server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java]: is it possible the taxmanager to be null ?

File: server/core/src/main/java/org/infinispan/server/core/logging/Log.java [hunk 5]
```diff
@@ -107,4 +107,7 @@
 
 //   @Message(value = "No provider for authorization realm", id = 5037)
```
Comment (karesti) [inline server/core/src/main/java/org/infinispan/server/core/logging/Log.java]: can't be delete this comment ?

File: server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java [hunk 6]
```diff
@@ -35,8 +35,18 @@ public String getType() {
    }
 
    public final T execute(TaskContext taskContext) {
-      Map<String, byte[]> rawParams = (Map<String, byte[]>) taskContext.getParameters().get();
-      Map<String, String> parameters = rawParams.entrySet().stream().collect(Collectors.toMap(Map.Entry::getKey, entry -> new String(entry.getValue(), CharsetUtil.UTF_8)));
+      Map<String, ?> raw = taskContext.getParameters().orElse(Collections.emptyMap());
```
Comment (karesti) [inline server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java]: For me as I said below, this is an anti-pattern. the TaskContext class should be refactored not to have the Optional<Map>

File: server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java [hunk 7]
```diff
@@ -35,8 +35,18 @@ public String getType() {
...
+      Map<String, String> parameters = raw.entrySet().stream().collect(
```
Comment (karesti) [inline server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java]: why not calling "map" method to transform

File: tasks/manager/src/test/java/org/infinispan/tasks/DummyTaskEngine.java (hunk 3 above)
Comment (tristantarrant) [reply tasks/manager/src/test/java/org/infinispan/tasks/DummyTaskEngine.java]: This is a test. Not bothering with making this perfect

File: server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java (hunk 4 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java]: Yes if the tasks library is missing. But then you cannot invoke any of the /tasks endpoints either

File: server/core/src/main/java/org/infinispan/server/core/logging/Log.java (hunk 5 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/logging/Log.java]: We leave unused logs commented in there to mark ids which should no longer be used

File: server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java (hunk 6 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java]: Not going to do this in 10.1.

File: server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java (hunk 7 above)
Comment (tristantarrant) [reply server/core/src/main/java/org/infinispan/server/core/admin/AdminServerTask.java]: I'm not going to do further refactorings to this code in this PR.

File: cli/cli-client/src/main/java/org/infinispan/cli/commands/Task.java [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+      @Override
+      public CommandResult exec(ContextAwareCommandInvocation invocation) {
+         if (help) {
+            invocation.println(invocation.getHelpInfo());
+         }
+         CommandInputLine cmd = new CommandInputLine(Task.CMD)
+               .arg(TYPE, Task.Exec.CMD)
+               .arg(NAME, taskName)
+               .arg(PARAMETERS, parameters != null ? parameters : Collections.emptyMap());
```
[the remark this answers is not in the document]
Comment (tristantarrant) [reply cli/cli-client/src/main/java/org/infinispan/cli/commands/Task.java]: Hey @karesti I tried doing that, but aesh resets it to null

File: client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+   }
+
+   @Override
+   public CompletionStage<RestResponse> list() {
+      return client.execute(baseURL);
+   }
+
+   @Override
+   public CompletionStage<RestResponse> exec(String taskName, Map<String, ?> parameters) {
+      Objects.nonNull(taskName);
```
Comment (pruivo) [inline client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java]: `Objects.requireNonNull(taskName);` ?

File: client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+      Objects.nonNull(parameters);
```
Comment (pruivo) [inline client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java]: `Objects.requireNonNull(parameters);` ?

File: cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java [hunk 11]
```diff
@@ -610,6 +633,13 @@ public String describeCounter(String container, String counter) throws IOExcepti
       return parseBody(fetch(() -> client.counter(counter).configuration()), String.class);
    }
 
+   @Override
+   public String describeTask(String container, String taskName) throws IOException {
+      List<Map<String, Object>> list = parseBody(fetch(() -> client.tasks().list()), List.class);
+      Optional<Map<String, Object>> task = list.stream().filter(i -> taskName.equals(i.get("name"))).findFirst();
+      return task.get().toString();
```
Comment (pruivo) [inline cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java]: `task.get()` will throw a `NoSuchElementException` if the task doesn't exist. 
is that what you want? wouldn't it be better to return an empty string?

File: core/src/main/java/org/infinispan/security/actions/SetCacheRebalancingAction.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+import org.infinispan.AdvancedCache;
+import org.infinispan.partitionhandling.AvailabilityMode;
+
+/**
+ * SetCacheAvailabilityAction.
+ *
+ * @author Dan Berindei
+ * @since 7.0
+ */
+public class SetCacheRebalancingAction extends AbstractAdvancedCacheAction<Void> {
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/security/actions/SetCacheRebalancingAction.java]: no sure why this class is here :)

File: server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+         NettyRestResponse.Builder builder = new NettyRestResponse.Builder();
+         try {
+            if (result instanceof byte[]) {
+               builder.contentType(TEXT_PLAIN_TYPE).entity(result);
+            } else {
+               byte[] resultBytes = invocationHelper.getMapper().writeValueAsBytes(result);
+               builder.contentType(APPLICATION_JSON_TYPE).entity(resultBytes);
+            }
+         } catch (JsonProcessingException e) {
+            builder.status(HttpResponseStatus.INTERNAL_SERVER_ERROR);
```
Comment (pruivo) [inline server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java]: don't you want to add the exception message to the response?

File: documentation/src/main/asciidoc/topics/rest_examples/get_v2_tasks_list.adoc [hunk 14]
```diff
@@ -0,0 +1 @@
+GET /rest/v2/tasks?action=list
```
Comment (pruivo) [inline documentation/src/main/asciidoc/topics/rest_examples/get_v2_tasks_list.adoc]: `action=list` isn't required. (if I understood the code correctly)

File: cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java (hunk 11 above)
Comment (tristantarrant) [reply cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java]: Changed to throw a Messages.noSuchResource() exception

File: client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java (hunk 10 above)
Comment (tristantarrant) [reply client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java]: Thanks, fixed.

File: client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java (hunk 9 above)
Comment (tristantarrant) [reply client/rest-client/src/main/java/org/infinispan/client/rest/impl/okhttp/RestTaskClientOkHttp.java]: Fixed, thanks

File: documentation/src/main/asciidoc/topics/rest_examples/get_v2_tasks_list.adoc (hunk 14 above)
Comment (tristantarrant) [reply documentation/src/main/asciidoc/topics/rest_examples/get_v2_tasks_list.adoc]: Fixed, thanks

File: server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java (hunk 13 above)
Comment (tristantarrant) [reply server/rest/src/main/java/org/infinispan/rest/resources/TasksResource.java]: Done, thanks

File: core/src/main/java/org/infinispan/security/actions/SetCacheRebalancingAction.java (hunk 12 above)
Comment (tristantarrant) [reply core/src/main/java/org/infinispan/security/actions/SetCacheRebalancingAction.java]: Removed it.

File: cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java [hunk 15]
```diff
@@ -50,8 +52,10 @@
 import org.infinispan.cli.commands.Site;
 import org.infinispan.cli.commands.Start;
 import org.infinispan.cli.commands.Stats;
+import org.infinispan.cli.commands.Task;
 import org.infinispan.cli.commands.Upgrade;
 import org.infinispan.cli.connection.Connection;
+import org.infinispan.cli.logging.Messages;
```
Comment (pruivo) [inline cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java]: unused import!

Comment (tristantarrant) [reply cli/cli-client/src/main/java/org/infinispan/cli/connection/rest/RestConnection.java]: Fixed

**Comment by tristantarrant** [comment]
Failures are unrelated

**Comment by pruivo** [comment]
@tristantarrant are you sure? 
` Unable to construct dependency org.infinispan.security.impl.AuthorizationHelper in scope GLOBAL for org.infinispan.scripting.ScriptingManager `

**Comment by tristantarrant** [comment]
Indeed. I've fixed it Let's wait for CI

**Comment by tristantarrant** [comment]
Failures are really unrelated now :)

**Pull request merged**